// @datadog/browser-rum/esm/rum.entry.js
import { BoundedBuffer, buildCookieOptions, checkCookiesAuthorized, checkIsNotLocalFile, createContextManager, deepClone, defineGlobal, getGlobalObject, isPercentage, makeGlobal, monitor, } from '@datadog/browser-core';
import { startRum } from './rum';
import { UserActionType } from './userActionCollection';
export var datadogRum = makeRumGlobal(startRum);
defineGlobal(getGlobalObject(), 'DD_RUM', datadogRum);
export function makeRumGlobal(startRumImpl) {
    var isAlreadyInitialized = false;
    var globalContextManager = createContextManager();
    var getInternalContextStrategy = function () {
        return undefined;
    };
    var beforeInitAddUserAction = new BoundedBuffer();
    var addUserActionStrategy = function (action) {
        beforeInitAddUserAction.add([action, deepClone(globalContextManager.get())]);
    };
    return makeGlobal({
        init: monitor(function (userConfiguration) {
            var _a;
            if (!checkCookiesAuthorized(buildCookieOptions(userConfiguration)) ||
                !checkIsNotLocalFile() ||
                !canInitRum(userConfiguration)) {
                return;
            }
            if (userConfiguration.publicApiKey) {
                userConfiguration.clientToken = userConfiguration.publicApiKey;
            }
            ;
            (_a = startRumImpl(userConfiguration, globalContextManager.get), getInternalContextStrategy = _a.getInternalContext, addUserActionStrategy = _a.addUserAction);
            beforeInitAddUserAction.drain(function (_a) {
                var action = _a[0], context = _a[1];
                return addUserActionStrategy(action, context);
            });
            isAlreadyInitialized = true;
        }),
        addRumGlobalContext: monitor(globalContextManager.add),
        removeRumGlobalContext: monitor(globalContextManager.remove),
        setRumGlobalContext: monitor(globalContextManager.set),
        getInternalContext: monitor(function (startTime) {
            return getInternalContextStrategy(startTime);
        }),
        addUserAction: monitor(function (name, context) {
            addUserActionStrategy({
                name: name,
                context: deepClone(context),
                startTime: performance.now(),
                type: UserActionType.CUSTOM,
            });
        }),
    });
    function canInitRum(userConfiguration) {
        if (isAlreadyInitialized) {
            if (!userConfiguration.silentMultipleInit) {
                console.error('DD_RUM is already initialized.');
            }
            return false;
        }
        if (!userConfiguration || (!userConfiguration.clientToken && !userConfiguration.publicApiKey)) {
            console.error('Client Token is not configured, we will not send any data.');
            return false;
        }
        if (!userConfiguration.applicationId) {
            console.error('Application ID is not configured, no RUM data will be collected.');
            return false;
        }
        if (userConfiguration.sampleRate !== undefined && !isPercentage(userConfiguration.sampleRate)) {
            console.error('Sample Rate should be a number between 0 and 100');
            return false;
        }
        if (userConfiguration.resourceSampleRate !== undefined && !isPercentage(userConfiguration.resourceSampleRate)) {
            console.error('Resource Sample Rate should be a number between 0 and 100');
            return false;
        }
        if (Array.isArray(userConfiguration.allowedTracingOrigins) &&
            userConfiguration.allowedTracingOrigins.length !== 0 &&
            userConfiguration.service === undefined) {
            console.error('Service need to be configured when tracing is enabled');
            return false;
        }
        return true;
    }
}
